### PR TITLE
I351 list results

### LINKED ISSUE #351: Find out for which model runs results are available
- Add a `--results` subcommand (or equivalent) to the `smif list` command which indicates which model runs have results and...
- whether the results are complete or not (i.e has been run successfully so that the results have a similar timestamp and all results files are present for all outputs listed in config)
- not that this raises the issue of having an incomplete implementation of a smif wrapper which fails to call the `set_results` method for a particular output, despite that output being listed in the config

Parent issue #350

### PR DESCRIPTION
**Pull request by fcooper8472**
#### `smif list`

`smif list` will now append an asterisk to model run names if full results exist

#### `smif available_results <run_name>`

`smif available_results <run_name>` will detail the results available for a single `<run_name>`: for instance (with results):
```
model run: energy_water_cp_cr
  - sos model: energy_water
    - sector model: water_supply
      - output: cost
        - decision 0: 2010, 2015, 2020
      - output: energy_demand
        - decision 0: 2010, 2015, 2020
      - output: water
        - decision 0: 2010, 2015, 2020
      - output: reservoir_level
        - decision 0: 2010, 2015, 2020
    - sector model: energy_demand
      - output: cost
        - decision 0: 2010, 2015, 2020
      - output: water_demand
        - decision 0: 2010, 2015, 2020
```
And, without results:
```
model run: energy_water_cp_cr
  - sos model: energy_water
    - sector model: water_supply
      - output: cost
        - no results
      - output: energy_demand
        - no results
      - output: water
        - no results
      - output: reservoir_level
        - no results
    - sector model: energy_demand
      - output: cost
        - no results
      - output: water_demand
        - no results
``` 

Omitting the `<run_name>` will print all information for every model run.

Closes #351 and #352

### CODE REVIEW

**Comment by fcooper8472** [comment]
@willu47 I have made the complete results checking optional: it's now an additional flag `smif list <-c/--complete>`.

Would be interesting to what is actually slowing up the smif list, but unfortunately I've not been able to get my dev version of smif working with nismod2...

I suspect the slowdown must be in `_get_canonical_expected_results` rather than `_get_canonical_available_results` but it only seems to be reading some YAML config - any ideas?

**Comment by tomalrussell** [comment]
I'm happy to ignore appveyor/branch failures here where they relate to conda-forge or dependency errors, so long as appveyor/pr and travis get :heavy_check_mark:s.

File: src/smif/cli/__init__.py [hunk 1]
```diff
@@ -104,8 +105,144 @@ def list_model_runs(args):
...
+    canonical_list = []
+
+    for t, d, sec_model_name, output_name in available_results:
+        canonical_list.append((t, 0, sec_model_name, output_name))
+
+    # Return as a set to remove duplicates
+    return set(canonical_list)
+
+
+def _list_available_results(store, run):
```
Comment (willu47) [inline src/smif/cli/__init__.py]: You may want to push this functionality into the store to enable #359 (or do that as part of #359), perhaps returning a dict which you then iterate over to print out the elements in this module?

File: src/smif/cli/__init__.py [hunk 2]
```diff
@@ -104,8 +105,144 @@ def list_model_runs(args):
...
+
+        expected_results = _get_canonical_expected_results(store, run_name)
+        available_results = _get_canonical_available_results(store, run_name)
+
+        complete = ' *' if expected_results == available_results else ''
+
+        print('{}{}'.format(run_name, complete))
+
+
+def _get_canonical_expected_results(store, model_run_name):
```
[the remark this answers is not in the document]
Comment (fcooper8472) [reply src/smif/cli/__init__.py]: Addressed in https://github.com/nismod/smif/pull/358/commits/8832224d0a87a5e9f924ecb6036966ab4a04def9

File: src/smif/cli/__init__.py [hunk 3]
```diff
@@ -104,8 +105,144 @@ def list_model_runs(args):
...
+        outputs = sec_model_config['outputs']
+
+        for output, t in itertools.product(outputs, timesteps):
+            expected_results.append((t, 0, sec_model_name, output['name']))
+
+    # Return as a set to remove duplicates
+    return set(expected_results)
+
+
+def _get_canonical_available_results(store, model_run_name):
```
[the remark this answers is not in the document]
Comment (fcooper8472) [reply src/smif/cli/__init__.py]: Addressed in https://github.com/nismod/smif/pull/358/commits/8832224d0a87a5e9f924ecb6036966ab4a04def9

File: src/smif/cli/__init__.py (hunk 1 above)
Comment (fcooper8472) [reply src/smif/cli/__init__.py]: This method has been refactored away completely.

**Comment by fcooper8472** [comment]
The new functionality builds the list of available results based on calls to `store.canonical_expected_results` and `store.available_results`.  This removes some code duplication and should also improve performance.

All methods (including new cli functionality) are now unit tested.

Ready for another review @willu47

**Pull request merged**